File: app/src/main/java/com/proxy/shadowsocksr/impl/crypto/crypto/CryptoManager.java
package com.proxy.shadowsocksr.impl.crypto.crypto;

import com.proxy.shadowsocksr.impl.crypto.CryptoUtils;

import java.nio.charset.Charset;
import java.util.HashMap;
import java.util.Map;

public final class CryptoManager
{
    private static CryptoManager cryptoManager = null;
    public static final Map<String, int[]> cipherList = new HashMap<>();
    //
    private static Map<String, byte[]> cachedKeys = new HashMap<>();

    private CryptoManager()
    {
    }

    public static CryptoManager getManager()
    {
        if (cryptoManager == null)
        {
            cryptoManager = new CryptoManager();
            cipherList.put("aes-128-cfb", new int[]{16, 16});
            cipherList.put("aes-192-cfb", new int[]{24, 16});
            cipherList.put("aes-256-cfb", new int[]{32, 16});
            cipherList.put("salsa20", new int[]{32, 8});
            cipherList.put("chacha20", new int[]{32, 8});
        }
        return cryptoManager;
    }

    public AbsCrypto getMatchCrypto(String cryptMethod, String pwd)
    {
        int[] cryptMethodInfo = getCipherInfo(cryptMethod);
        //
        String k = cryptMethod + ":" + pwd;
        byte[] key;
        if (cachedKeys.containsKey(k))
        {
            key = cachedKeys.get(k);
        }
        else
        {
            byte[] passbf = pwd.getBytes(Charset.forName("UTF-8"));
            key = new byte[cryptMethodInfo[0]];
            CryptoUtils.EVP_BytesToKey(passbf, key);
            cachedKeys.put(k, key);
        }
        //
        if (cryptMethod.startsWith("aes-"))
        {
            return new AESCrypto(cryptMethod, key);
        }
        else if (cryptMethod.equals("salsa20"))
        {
            return new SalsaCrypto(cryptMethod, key);
        }
        //....
        else
        {
            return new ChachaCrypto(cryptMethod, key);
        }
    }

    public int[] getCipherInfo(String name)
    {
        if (cipherList.containsKey(name))
        {
            return cipherList.get(name);
        }
        return new int[]{32, 8};
    }
}

File: app/src/main/java/com/proxy/shadowsocksr/impl/crypto/CryptoManager.kt
package com.proxy.shadowsocksr.impl.crypto

import com.proxy.shadowsocksr.impl.ImplUtils
import com.proxy.shadowsocksr.impl.crypto.crypto.AbsCrypto
import com.proxy.shadowsocksr.impl.crypto.crypto.AesCFBCrypto
import com.proxy.shadowsocksr.impl.crypto.crypto.ChachaCrypto
import com.proxy.shadowsocksr.impl.crypto.crypto.SalsaCrypto
import java.nio.charset.Charset

object CryptoManager
{
    private val cipherList = mapOf(
            Pair("aes-128-cfb", intArrayOf(16, 16)),
            Pair("aes-192-cfb", intArrayOf(24, 16)),
            Pair("aes-256-cfb", intArrayOf(32, 16)),
            Pair("salsa20", intArrayOf(32, 8)),
            Pair("chacha20", intArrayOf(32, 8)))

    private val cachedKeys: Map<String, ByteArray> = mapOf()

    fun getMatchCrypto(cryptMethod: String = "chacha20", password: String = "0"): AbsCrypto
    {
        val cryptMethodInfo:IntArray = getCipherInfo(cryptMethod)
        val k = "$cryptMethod:$password"
        var key: ByteArray
        if (cachedKeys.containsKey(k))
        {
            key = cachedKeys.getOrImplicitDefault(k)
        }
        else
        {
            val passbf = password.toByteArray(Charset.forName("UTF-8"))
            key = ByteArray(cryptMethodInfo[0])
            ImplUtils.EVP_BytesToKey(passbf, key)
            cachedKeys.plus(Pair(k, key))
        }
        //
        if (cryptMethod.startsWith("aes-"))
        {
            return AesCFBCrypto(cryptMethod, key)
        }
        //...
        when (cryptMethod)
        {
            "salsa20" -> return SalsaCrypto(cryptMethod, key)
        //...
            else -> return ChachaCrypto(cryptMethod, key)//default chacha20
        }
    }

    fun getCipherInfo(name: String): IntArray
    {
        return cipherList.getOrImplicitDefault(name)
    }
}
